### ISSUE TITLE
Allow `optional` in tuble types

### ISSUE THREAD

**Issue by jquesada2016** [edgedb/edgedb#5778]
I believe it would be a fantastic idea if we could mark tuple fields as optional. Something like:

```
tuple<
  name: str,
  optional_data: optional str
>
```

This would remove 90% of the use of `json`, especially for query parameters, as currently, there's no way to model optional data in parameters without the use of `json`.

I can imagine this feature posing problems if allowed inside unnamed tuples, such as `tuple<str, optional str, str>`, as ambiguity will arise when parsing. I'll leave this problem as an exercise to the team, if such a feature would be desirable (: (maybe eager parsing?)

Alternatively, one can currently model optional data pairing `tuple` with `json`, for example:

```
tuple<
  name: str,
  optional_data: json
>
```

which can later be used like:

```
<str>data.optional_data
```

Nevertheless, even though it _can_ be currently done, supporting `optional` inside tuple types would almost complete data-modeling capabilities for query parameters especially.

**Comment by aljazerzen** [comment]
This sounds like a reasonable request, but unfortunately I think it clashes with semantics of optional and empty sets.

When you write:
```
select (A, B)
```
... where `A` and `B` are some object types (or anything else really), this will produce Cartesian product of all values of `A` and all values of `B`.

For example:
```
> select ({1, 2, 3}, {true, false})
{(1, true), (1, false), (2, true), (2, false), (3, true), (3, false)}
```

Now, optional means (in this scenario) that the value can either have cardinality 1 or 0. This is problematic, because a tuple with a field of cardinality 0 would produce:
```
> select ({1, 2, 3}, {}, {true, false})
{}
```
... an empty set.

---

I'm not sure if we support this yet, but this might be possible using free objects instead of tuples. @vpetrovykh would know more about this.

**Comment by elprans** [comment]
Allowing free objects as input is, indeed, the way to go.  We need to implement [typing free objects](https://github.com/edgedb/rfcs/blob/master/text/1022-freetypes.rst#typing-free-objects) from RFC 1022 first.  Other prerequisites are in place, I think.

**Comment by AnsonHwang86** [comment]
The rfcs document said that `Typing free objects sounds like an EdgeDB 4.0 feature, especially given that we enable tuples to be used as query arguments in 3.0.`. 
I just want to confirm if free object will including in the coming v4?

**Comment by aljazerzen** [comment]
Unfortunately no, user-defined free objects will not be included in v4. This is partially due to increased release cadence, which means that we ship more often than before.

**Comment by jackfischer** [comment]
A hearty +1 on the free objects alternative. Our team does lots of bulk queries and are finding ourselves either polluting our data with null object pattern type noise or introducing ternaries inside for loops to individually swap out null objects for proper edgedb empty sets. Worrying about this is a notable extra layer of complexity for taking advantage of edgedb.

**Comment by aljazerzen** [comment]
> say that you can have an optional element in a named tuple

This is not true (or I'm misunderstanding). Can you provide demonstration + link to docs that says that?

**Comment by jquesada2016** [comment]
Any updates on this?

**Comment by aljazerzen** [comment]
@MiroslavPetrik when a tuple contains sets, the sets are "lifed trough the tuple" to the top-level via cartesian multiplication.

For example:
```
> select (a := 1, b := {true, false})
{(1, true), (1, false)}
```

From which follows:

```
> select (a := 1, b := {})
{}
```

So this is expected behavior, although quite unconventional.

Having `optional` in tuple, as you propose, would make it behave as you want.
